core/src/main/java/org/frankframework/pipes/ReplacerPipe.java

```diff
@@ -70,24 +95,67 @@ public void configure() throws ConfigurationException {
 
 	@Override
 	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
-		ReplacingInputStream inputStream = getReplacingInputStream(message, getFind(), getReplace());
+		try {
+			// Make sure to close the incoming message
+			message.unscheduleFromCloseOnExitOf(session);
 
-		Message result = new Message(inputStream, message.copyContext().withoutSize());
-		// As we wrap the input-stream, we should make sure it's not closed when the session is closed as that might close this stream before reading it.
-		message.unscheduleFromCloseOnExitOf(session);
-		result.closeOnCloseOf(session, this);
+			// Create a ReplacingInputStream for find/replace
+			ReplacingInputStream replacingInputStream = new ReplacingInputStream(message.asInputStream(), find, replace, isReplaceNonXmlChars(),
+					getNonXmlReplacementCharacter(), isAllowUnicodeSupplementaryCharacters()
+			);
 
-		return new PipeRunResult(getSuccessForward(), result);
-	}
+			// Replace parameters
+			ReplacingVariablesInputStream replaceParametersStream = new ReplacingVariablesInputStream(replacingInputStream, "?",
+					getKeyValueMapForParameters(message, session)
+			);
 
-	private ReplacingInputStream getReplacingInputStream(Message message, String find, String replace) throws PipeRunException {
-		try {
-			return new ReplacingInputStream(message.asInputStream(), find, replace, isReplaceNonXmlChars(), getNonXmlReplacementCharacter(), isAllowUnicodeSupplementaryCharacters());
+			// Wrap for 'substitute vars' if necessary
+			ReplacingVariablesInputStream inputStream = wrapWithSubstituteVarsInputStreamIfNeeded(replaceParametersStream);
+
+			Message result = new Message(inputStream);
+			result.closeOnCloseOf(session, this);
+
+			return new PipeRunResult(getSuccessForward(), result);
 		} catch (IOException e) {
 			throw new PipeRunException(this, "cannot open stream", e);
 		}
 	}
 
+	/**
+	 * If subsituteVars is true, we need to wrap the inputStream again to substitute ${} syntax variables with
+	 * system properties, session variables and application properties.
+	 */
+	private ReplacingVariablesInputStream wrapWithSubstituteVarsInputStreamIfNeeded(ReplacingVariablesInputStream replaceParametersStream) {
+		if (substituteVars) {
+			return new ReplacingVariablesInputStream(replaceParametersStream, "$", getKeyValueMapForAppConstants());
+		}
+
+		return replaceParametersStream;
+	}
+
+	/**
+	 * @return the appConstants as a key/value map combined with the session values
+	 */
+	private Map<String, String> getKeyValueMapForAppConstants() {
+		return appConstants.entrySet().stream()
+				.collect(Collectors.toMap(entry -> entry.getKey().toString(), entry -> entry.getValue().toString()));
```

Comment:
Dit gaat helaas fout, properties kunnen verwijzen naar andere properties, (of een samenstelling zijn van meerdere properties). Nu pakt hij de letterlijkt (un-resolved) waardes.